File: model/lecture.js
var mongoose = require('../db');

var LectureSchema = mongoose.Schema({
  year: { type: Number, required: true },           // 연도
  semester: { type: Number, required: true },       // 학기
  classification: { type: String, required: true }, // 교과 구분
  department: String,                               // 학부
  academic_year: String,                            // 학년
  course_number: { type: String, required: true },  // 교과목 번호
  lecture_number: String,                           // 강좌 번호
  course_title: { type: String, required: true },   // 과목명
  credit: Number,                                   // 학점
  class_time: String,
  class_time_json: [
  { day : Number, start: Number, len: Number, place : String }
  ],
  class_time_mask: { type: [ Number ], required: true },
  instructor: String,                               // 강사
  quota: Number,                                    // 정원
  enrollment: Number,                               // 신청인원
  remark: String,                                   // 비고
  category: String,
  created_at: Date,
  updated_at: Date
});

/*
 * Lecture.add_lecture(lecture)
 * 연도, 학기, 교과목 번호와 강좌 번호를 비교하여 같은 강좌인지 판단.
 * param =======================================
 * lecture : target for comparison
 */
LectureSchema.methods.is_equal = function(lecture) {
  return (this.year == lecture.year &&
      this.semester == lecture.semester &&
      this.course_number == lecture.course_number &&
      this.lecture_number == lecture.lecture_number);
};

LectureSchema.statics.is_equal = function(lecture1, lecture2) {
  return (lecture1.year == lecture2.year &&
  lecture1.semester == lecture2.semester &&
  lecture1.course_number == lecture2.course_number &&
  lecture1.lecture_number == lecture2.lecture_number);
};

LectureSchema.index({ year: 1, semester: 1, classification: 1 });
LectureSchema.index({ year: 1, semester: 1, department: 1 });
LectureSchema.index({ year: 1, semester: 1, course_title: 1 });

module.exports = mongoose.model('Lecture', LectureSchema);

File: core/src/main/kotlin/lectures/data/Lecture.kt
package com.wafflestudio.snutt.lectures.data

import com.fasterxml.jackson.annotation.JsonProperty
import com.wafflestudio.snutt.common.enums.Semester
import org.springframework.data.annotation.Id
import org.springframework.data.mongodb.core.index.CompoundIndex
import org.springframework.data.mongodb.core.mapping.Document
import org.springframework.data.mongodb.core.mapping.Field

@Document("lectures")
@CompoundIndex(def = "{ 'year': 1, 'semester': 1 }")
@CompoundIndex(def = "{ 'course_number': 1, 'lecture_number': 1 }")
data class Lecture(
    @Id
    @param:JsonProperty("_id")
    var id: String? = null,
    @Field("academic_year")
    var academicYear: String?,
    var category: String?,
    @Field("class_time_json")
    var classPlaceAndTimes: List<ClassPlaceAndTime>,
    var classification: String?,
    var credit: Long,
    var department: String?,
    var instructor: String?,
    @Field("lecture_number")
    var lectureNumber: String,
    var quota: Int,
    var freshmanQuota: Int? = null,
    var remark: String?,
    var semester: Semester,
    var year: Int,
    @Field("course_number")
    var courseNumber: String,
    @Field("course_title")
    var courseTitle: String,
    var registrationCount: Int = 0,
    var wasFull: Boolean = false,
    var evInfo: EvInfo? = null,
    var categoryPre2025: String?,
    @Field("academic_year_en")
    var academicYearEn: String? = null,
    @Field("category_en")
    var categoryEn: String? = null,
    @Field("classification_en")
    var classificationEn: String? = null,
    @Field("department_en")
    var departmentEn: String? = null,
    @Field("instructor_en")
    var instructorEn: String? = null,
    @Field("remark_en")
    var remarkEn: String? = null,
    @Field("course_title_en")
    var courseTitleEn: String? = null,
) {
    infix fun equalsMetadata(other: Lecture): Boolean =
        this === other ||
            (
                academicYear == other.academicYear &&
                    category == other.category &&
                    classPlaceAndTimes == other.classPlaceAndTimes &&
                    classification == other.classification &&
                    credit == other.credit &&
                    department == other.department &&
                    instructor == other.instructor &&
                    lectureNumber == other.lectureNumber &&
                    quota == other.quota &&
                    freshmanQuota == other.freshmanQuota &&
                    remark == other.remark &&
                    semester == other.semester &&
                    year == other.year &&
                    courseNumber == other.courseNumber &&
                    courseTitle == other.courseTitle &&
                    categoryPre2025 == other.categoryPre2025 &&
                    academicYearEn == other.academicYearEn &&
                    categoryEn == other.categoryEn &&
                    classificationEn == other.classificationEn &&
                    departmentEn == other.departmentEn &&
                    instructorEn == other.instructorEn &&
                    remarkEn == other.remarkEn &&
                    courseTitleEn == other.courseTitleEn
            )
}
